clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -203,6 +204,7 @@ public String protocolType() {
 
         for (PartitionAssignor assignor : assignors) {
             Subscription subscription = assignor.subscription(joinedSubscription);
+            subscription.setOwnedPartitions(new ArrayList<>(subscriptions.assignedPartitions()));
```

Comment:
I'm going back and forth on this but actually I think it's best to push this into the Assignor as well, and not *just* because it's more symmetric. I see two arguments for doing it this way:

1) the PartitionAssignor doesn't necessarily know anything about `ownedPartitions`
2) even if it does, it will not know about updates (such as metadata/subscription change, missed rebalance) while the ConsumerCoordinator's `subscriptions.assignedPartitions` is always up to date.

To address 1), I think it is reasonable to assume/force a **Cooperative** assignor to have some knowledge of `ownedPartitions` especially since we are ok to assume it will be able to handle the "adjust assignment" logic. And regarding 2), we could actually just pass along the current assignment to the PartitionAssignor prior, or as a separate `PartitionAssignor.cooperativeSubscription(topics, assignment)` method that is called in cooperative mode instead of the above. 

One upside of setting ownedPartitions in ConsumerCoordinator is that the PartitionAssignor no longer needs to know/save its own assignment. However what we lose here is the ability for the PartitionAssignor to react to anything other than the currently owned partitions.Imagine we have a heavily stateful app and a consumer misses a rebalance. It's partitions get assigned to another consumer but there's so much state to restore it doesn't get very far before the first consumer manages to rejoin. The first consumer's state is up to date while the second consumer is very far behind, but only the second consumer will join with a non-empty `ownedPartitions` so it is more likely to get the partitions again.

(this scenario requires a more advanced assignment strategy than we currently employ, but even the current sticky assignor leverages this information to improve stickiness where possible)  @guozhangwang @hachikuji

Reply:
After discussing this offline I agree to stick with just letting the ConsumerCoordinator set ownedPartitions and remove userdata completely for cooperative sticky assignor